### PR TITLE
airsonic: force use of jre8

### LINKED ISSUE #121556: Airsonic fails to start on latest unstable
**Describe the bug**

The Airsonic service fails to start.

**To Reproduce**
1. Add

        services.airsonic.enable = true;

   to your config.

2. `sudo nixos-rebuild switch --upgrade`
3. `journalctl -xefu airsonic` reveals the following error:

        java.lang.IllegalStateException: Cannot load configuration class: org.airsonic.player.Application

    which causes the application to enter a restart loop.

<details>
<summary>Error trace</summary>
<p>

```
May 02 21:19:56 coruscant java[1518405]: 2021-05-02 21:19:56.987 ERROR --- o.s.boot.SpringApplication               : Application startup failed
May 02 21:19:56 coruscant java[1518405]: java.lang.IllegalStateException: Cannot load configuration class: org.airsonic.player.Application
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.annotation.ConfigurationClassPostProcessor.enhanceConfigurationClasses(ConfigurationClassPostProcessor.java:404) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanFactory(ConfigurationClassPostProcessor.java:249) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:283) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:127) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:687) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:524) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:123) ~[spring-boot-1.5.22.RELEASE.jar!/:1.5.22.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:666) ~[spring-boot-1.5.22.RELEASE.jar!/:1.5.22.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:353) ~[spring-boot-1.5.22.RELEASE.jar!/:1.5.22.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.SpringApplication.run(SpringApplication.java:300) ~[spring-boot-1.5.22.RELEASE.jar!/:1.5.22.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:135) ~[spring-boot-1.5.22.RELEASE.jar!/:1.5.22.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.airsonic.player.Application.main(Application.java:207) ~[classes!/:10.6.2-RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at java.base/java.lang.reflect.Method.invoke(Method.java:567) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) ~[airsonic.war:10.6.2-RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) ~[airsonic.war:10.6.2-RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.loader.Launcher.launch(Launcher.java:50) ~[airsonic.war:10.6.2-RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.boot.loader.WarLauncher.main(WarLauncher.java:59) ~[airsonic.war:10.6.2-RELEASE]
May 02 21:19:56 coruscant java[1518405]: Caused by: java.lang.ExceptionInInitializerError: null
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.annotation.ConfigurationClassEnhancer.newEnhancer(ConfigurationClassEnhancer.java:122) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.annotation.ConfigurationClassEnhancer.enhance(ConfigurationClassEnhancer.java:110) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.context.annotation.ConfigurationClassPostProcessor.enhanceConfigurationClasses(ConfigurationClassPostProcessor.java:394) ~[spring-context-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         ... 19 common frames omitted
May 02 21:19:56 coruscant java[1518405]: Caused by: org.springframework.cglib.core.CodeGenerationException: java.lang.reflect.InaccessibleObjectException-->Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @7bfcd12c
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.ReflectUtils.defineClass(ReflectUtils.java:464) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:336) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:93) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:91) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.internal.LoadingCache$2.call(LoadingCache.java:54) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.internal.LoadingCache.createEntry(LoadingCache.java:61) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.internal.LoadingCache.get(LoadingCache.java:34) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData.get(AbstractClassGenerator.java:116) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:291) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.KeyFactory$Generator.create(KeyFactory.java:221) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.KeyFactory.create(KeyFactory.java:174) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.KeyFactory.create(KeyFactory.java:153) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.proxy.Enhancer.<clinit>(Enhancer.java:73) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         ... 22 common frames omitted
May 02 21:19:56 coruscant java[1518405]: Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @7bfcd12c
May 02 21:19:56 coruscant java[1518405]:         at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at java.base/java.lang.reflect.Method.setAccessible(Method.java:193) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.ReflectUtils$1.run(ReflectUtils.java:61) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at java.base/java.security.AccessController.doPrivileged(AccessController.java:554) ~[na:na]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.ReflectUtils.<clinit>(ReflectUtils.java:52) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.KeyFactory$Generator.generateClass(KeyFactory.java:243) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         at org.springframework.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:329) ~[spring-core-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
May 02 21:19:56 coruscant java[1518405]:         ... 34 common frames omitted
```

</p>
</details>

**Expected behavior**
It doesn't error, and instead runs successfully.

**Additional context**
It doesn't seem to be due to an Airsonic update that this started to fail, so my guess is that something got changed with the JRE.

I think Airsonic will only work with JDK 8 (see: https://airsonic.github.io/docs/install/war-standalone/), but I'm not very familiar with how Java works, so I'm not sure.

**Notify maintainers**

@disassembler 

**Metadata**
Fails on both of the following systems:
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.30, NixOS, 21.05pre285574.8e4fe32876c (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210326_dd77f71`
 - channels(root): `"nixos-21.05pre285574.8e4fe32876c"`
 - channels(sumner): `"home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.107, NixOS, 21.05pre286571.c576998594b (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-21.05pre286571.c576998594b"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
  airsonic
```

### PR DESCRIPTION
**Pull request by sumnerevans**
###### Motivation for this change

Fixes #121556. This forces the airsonic service to use `jre8` instead of the `jre` package.

Also adds a test to verify that the airsonic service is working.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by lukegb — APPROVED** [review]

**Comment by lukegb** [comment]
It would be nice if this module had a test, heh.

**Comment by sumnerevans** [comment]
@lukebg I added a test for the module. It fails without the fix for jre8, and succeeds with the fix. The one concern I have is that I don't know how to make the test time out because the systemd service just boot-loops in the unfixed case.

**Comment by lukegb** [comment]
> @lukegb I added a test for the module. It fails without the fix for jre8, and succeeds with the fix. The one concern I have is that I don't know how to make the test time out because the systemd service just boot-loops in the unfixed case.

Oh nice! Thanks for doing that :)

On Hydra, it'll time out eventually. OfBorg has a build timeout eventually as well, so that's probably fine.

Can you please add your test to the airsonic package's passthru.tests attribute (https://nixos.org/manual/nixpkgs/unstable/#var-meta-tests)? Thanks!

**Comment by sumnerevans** [comment]
@lukegb hmm... looks like it didn't actually work... from what I can tell the test "works on my machine" haha. Any ideas? Maybe we skip adding the test for now and just get the fix in?

File: nixos/tests/airsonic.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    machine.wait_for_unit("airsonic.service")
+    machine.wait_for_open_port(4040)
+
+    with machine.nested("Waiting for UI to work"):
+        retry(airsonic_is_up)
+  '';
+
+  # Airsonic is a Java application, and unfortunately requires a significant
+  # amount of memory.
+  virtualisation.memorySize = 1024;
```
Comment (lukegb) [inline nixos/tests/airsonic.nix:28]: This, somewhat counter-intuitively, needs to be in the machine description.

You probably also want to set services.airsonic.maxMemory = 800;

Then the test passes for me in about a minute and a half without too much GC thrashing.

Comment (sumnerevans) [reply nixos/tests/airsonic.nix]: Oh, that makes sense. Fixed. 🤞 that the hydra build works.

**Comment by lukegb** [comment]
@ofborg build airsonic airsonic.tests

**Comment by sumnerevans** [comment]
It looks like it failed on darwin, which makes sense because it uses systemd. Any advice on how to fix that?

**Comment by lukegb** [comment]
That's kind of an expected failure - we don't generally expect NixOS tests to work properly on darwin (which is why it's marked as a success anyway)

**Comment by lukegb** [comment]
Thanks! :)

**Pull request merged**